### PR TITLE
Use gold instead of the traditional linker if possible.

### PR DESCRIPTION
**Pull request by tamiko**
Supersedes #1492

### CODE REVIEW

**Comment by tamiko** [comment]
@bangerth @tjhei This is the old pull-request with an additional commit that refactor the check into ```check_01_compiler_features.cmake```. This should address all problems.

File: cmake/checks/check_01_compiler_features.cmake [hunk 1]
```diff
@@ -336,3 +337,26 @@ CHECK_CXX_SOURCE_COMPILES(
...
+#
+# We have to try to link a full executable with -fuse-ld=gold to check
+# whether "ld.gold" is actually available. gcc has the bad habit of
+# accepting the flag without emitting an error.
+#
+# Wolfgang Bangerth, Matthias Maier, 2015
+#
+PUSH_CMAKE_REQUIRED("-Werror")
+PUSH_CMAKE_REQUIRED("-fuse-ld=gold")
+CHECK_CXX_SOURCE_COMPILES(
```
Comment (bangerth) [inline cmake/checks/check_01_compiler_features.cmake]: I think the best solution would actually be to ensure that it doesn't just compile but also links. Is this implied?

Comment (tamiko) [inline cmake/checks/check_01_compiler_features.cmake]: > I think the best solution would actually be to ensure that it doesn't
> just compile but also links. Is this implied?

Yes, this is implied. The macro name is one of cmake's glorious
mistakes.

Comment (bangerth) [inline cmake/checks/check_01_compiler_features.cmake]: OK, thanks!

**Pull request merged**